### PR TITLE
Initial implementation of upgrade logic for apt based OS (Ubuntu)

### PR DESCRIPTION
**Pull request by tobru**
This script implements the logic for upgrading packages on APT based Linux distributions - mainly Ubuntu

### CODE REVIEW

File: main.sh [hunk 1]
```diff
@@ -362,32 +364,104 @@ fi
...
 [[ "${LOG_LEVEL:-}" ]] || emergency "Cannot continue without LOG_LEVEL. "
 
 
+### Package manager functions
+##############################################################################
+
+function run_apt_get_update() {
+  # Disable errexit as we want to catch the error to display
+  # some informative log messages
+  set +e
```
Comment (simu) [inline main.sh]: This pattern is unnecessary, change the function to the following:

```
function run_apt_get_update() {
  if ! apt_get=$(DEBIAN_FRONTEND=noninteractive apt-get -q update 2>&1); then
    error "${apt_get}"
    emergency "Package cache refresh failed"
  fi

  info "${apt_get}"
  info "Package cache refresh finished"
}
```

File: main.sh [hunk 2]
```diff
@@ -362,32 +364,104 @@ fi
...
+  info "Package cache refresh finished"
+
+  # Enable errexit again
+  set -e
+}
+
+function run_apt_get_upgrade() {
+  # Disable errexit as we want to catch the error to display
+  # some informative log messages
+  set +e
```
Comment (simu) [inline main.sh]: Apply same pattern as for `run_apt_get_update`.

**Review by simu — DISMISSED** [review]
Overall the implementation looks good, the trick with `set +e` and `set -e` to collect error output from `apt-get` is unnecessary, see inline comments

File: main.sh (hunk 1 above)
Comment (mhutter) [reply main.sh]: IIRC `set -e` does not even propagate inside functions

Comment (simu) [reply main.sh]: According to my local testing `set -e` does in fact apply to code inside functions

**Comment by mhutter** [comment]
Moving all the helper functions (ie 90% of the code) to a separate file and sourcing it from the main script would greatly increase legibility

**Comment by tobru** [comment]
@mhutter If it's OK for you I would do the optimization in a follow-up PR.

**Comment by tobru** [comment]
After giving it some more thought, this is way over-engineered for the time being. It should really just work in a busybox with `/bin/sh`. Most of the boilerplate functionality is anyways not needed. I'll close this issue for good and give it another try.

**Pull request closed without merging**